src/parquet/schema.cc

```diff
@@ -620,6 +620,15 @@ const ColumnDescriptor* SchemaDescriptor::Column(int i) const {
   return &leaves_[i];
 }
 
+int SchemaDescriptor::ColumnIndex(const NodePtr& node) const {
```

Comment:
>And imho it seems that what matters anyways in terms of unordered_map is that they are hashable and comparable which should be defined on the object and not the shared_ptr.

I understand that, but it requires we supply our Hash and Predict class which adds complexity.
>Another option is to use std::string for the key in conjunction with ColumnPath::ToDotString. But I'm guessing this will practically be slower than the O(n) array pass because of all the string manipulation.

This is what I want to proposal.  Didn't make it clear in the previous comment.
As for performance, It might be slower for small size of leaf columns in the schema.  But it's bounded to O(1) look up if we be careful and cache ColumnPath::ToDotString.

It's also easy for user to use this api, they can just supply a string and query.